### PR TITLE
feat(storage): introduce dedup_pk_row encoding for StateTable

### PR DESCRIPTION
**Pull request by kwannoel**
## What's changed and what's your intention?

- TODOs
  - [x] Add `DedupPkStateTable`.
  - [x] test virtual inner state table enc/dec against dedup pk cell table.
  - [x] implement encode / decode logic.
  - [x] test in test_relational_table
  - [x] test memcomparable
  - [ ] test backwards (read) compatibility with normal rows?
  - [ ] use `DedupPkStateTable` for mview.
  - [ ] update locations which use state table `get_row` or iterators.
  - [ ] ensure e2e tests cover interactions between read and writes for dedup encoding.
- Summarize your change (**mandatory**)
  - Introduces `DedupPkStateTable`. It has the same interface as StateTable, but also deduplicates pk datums from rows.
- How does this PR work? Need a brief introduction for the changed logic (optional)
  - Internally it uses `StateTable` to interface with Storage.
  - `StateTable` interface is extended with 3 more `iter` methods. These are counterparts to existing iter methods. They yield `keys` as well.
  - `DedupStateTable` uses these `iter` methods to retrieve keys for decoding dedup pk rows to normal rows.
- Describe clearly one logical change and avoid lazy messages (optional)
  - DedupPkStateTable needs to follow dedup pk encoding on the cell level. In this means that given the following rows:
    ```rs
    let pk_index = 1;
    let column_ids = vec![0, 1, 2];
    let row_1 = vec!["a", "b", "c"];
    let row_2 = vec!["d", "e", "f"]];
    ```
    We should have the following in cell based storage:
    ```
    row_1: "b" 0 -> "a", "b" 2 -> "c"
    row_2: "e" 0 -> "d", "e" 2 -> "f"
    ```
  This encoding can be reproduced by:
  1.  `StateTable` is instantiated with `partial_column_descs`. These are the column_descs of dedup pk row datums.
  2. We insert rows with pk datums filtered out.
- Describe any limitations of the current code (optional)
  - Memcomparable pks are still duplicated.
  - If more iter methods are introduced to StateTable, they would all need to be duplicated to yield keys.
  - Decoding has higher overhead in `DedupPkStateTable` than `DedupPkCellBasedRowIter`.
  - https://github.com/singularity-data/risingwave/blob/4ac214cd5ba8063054d2d4659e187d3d3814e393/src/storage/src/table/dedup_pk_state_table.rs#L134-L151
  - https://github.com/singularity-data/risingwave/blob/3bca26e56247d1a2dea1d588abfc84ec5a15cf3e/src/storage/src/table/cell_based_table.rs#L631-L639
- Other designs considered
  Considered if I should parameterize on cellbased ser/de instead, similar to how `StateStore` is parameterized. However that would require changes the semantics of cellbased ser/de from row. It is not a transform from cells -> row, now it is `cells X pk -> row`. It will also change interface of State and CellBasedTable, since now `cellbased ser/de` is a parameter.

## Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

#588

### CODE REVIEW

**Comment by kwannoel** [comment]
Currently thinking about design further:

- pk is required to decode dedup pk rows to normal rows.
- since `DedupPkStateTable` is a wrapper around `StateTable`, it should wrap `StateTable` api (`iter_with_pk_prefix`, `iter`, `iter_with_pk_bounds`) to implement its own set of iterators.
- However these iterators only yield rows, without pk. Meaning we can't decode.
- This means I will need to introduce 3 more methods (one for each `StateTable` iterator method) which yield keys as well. In `DedupPkStateTable` I can then call these methods and decode with key.
- This gets the jobs done but I'm wondering if there's a better idea.

**Comment by kwannoel** [comment]
Updated with these changes.

> Currently thinking about design further:
> 
>     * pk is required to decode dedup pk rows to normal rows.
> 
>     * since `DedupPkStateTable` is a wrapper around `StateTable`, it should wrap `StateTable` api (`iter_with_pk_prefix`, `iter`, `iter_with_pk_bounds`) to implement its own set of iterators.
> 
>     * However these iterators only yield rows, without pk. Meaning we can't decode.
> 
>     * This means I will need to introduce 3 more methods in `StateTable` (one for each `StateTable` iterator method) which yield keys as well. In `DedupPkStateTable` I can then call these methods and decode with key.
> 
>     * This gets the job done but I'm wondering if there's a better idea.

**Comment by kwannoel** [comment]
Need to write a test to ensure the following:
- Its encoding needs to match what dedup pk cell based row iter expects.
- CellBasedTableRowIter uses `CellBasedRowDeserializer` to deserialize.
- `CellBasedRowDeserializer` allocates a `Row`, of size `column_len` (which is the normal row size without dedup encoding).
- This can be tested with `StateTable` (used by `DedupPkStateTable` under the hood) and `DedupPkCellBasedTable` in `test_relational_table`.

File: src/storage/src/table/state_table.rs [hunk 1]
```diff
@@ -240,10 +240,49 @@ impl<S: StateStore> StateTable<S> {
...
+            .await?;
+        Ok(stream.map(|r| r.map(|(_k, v)| v)))
+    }
 }
 
 pub trait RowStream<'a> = Stream<Item = StorageResult<Cow<'a, Row>>>;
 
+pub type RawKey = Vec<u8>;
+pub trait KeyAndRowStream<'a> =
+    StreamExt<Item = StorageResult<(Cow<'a, RawKey>, Cow<'a, Row>)>> + 'a;
```
Comment (skyzh) [inline src/storage/src/table/state_table.rs:284]: StreamExt -> Stream?

**Review by skyzh — COMMENTED** [review]
> these iterators only yield rows, without pk

I remembered we have an iterator returning pk inside StateStable. We can modify the state table API to have some interfaces return key + decoded value.

> This means I will need to introduce 3 more methods in StateTable (one for each StateTable iterator method) which yield keys as well

This is good enough!

**Comment by kwannoel** [comment]
> Need to write a test to ensure the following:
> 
>     * Its encoding needs to match what dedup pk cell based row iter expects.
> 
>     * CellBasedTableRowIter uses `CellBasedRowDeserializer` to deserialize.
> 
>     * `CellBasedRowDeserializer` allocates a `Row`, of size `column_len` (which is the normal row size without dedup encoding).
> 
>     * This can be tested with `StateTable` (used by `DedupPkStateTable` under the hood) and `DedupPkCellBasedTable` in `test_relational_table`.

Latest commits include these tests, shows the current approach is feasible.

**Comment by kwannoel** [comment]
Done with `DedupPkTable` implementation.
Next step will be to refactor `mview` to use it, to make sure it works e2e as well.

**Comment by kwannoel** [comment]
Overlooked `distribution_keys`, used by `batch_write_rows_inner`. It tries to hash on de-duplicated pk datums. Design probably needs to change, hopefully not too much.

**Comment by kwannoel** [comment]
> Overlooked `distribution_keys`, used by `CellBasedTable::batch_write_rows_inner`. It tries to hash on de-duplicated pk datums. Design probably needs to change, hopefully not too much.

Off the top of my head:
1. hash with key instead? This means we need to change hasher to be on slice instead :thinking: . Can we just hash the whole vec? Probably should read some docs on hashing design.
2. compute hash earlier when pk still exists as row and pass it downstream?

**Comment by skyzh** [comment]
The 3rd approach looks good to me. We can let the serializer to do the serialization while calculating hash. Note that hash is calculated with distribution key instead of the full pk.

**Comment by kwannoel** [comment]
Pushed failing commit: https://github.com/singularity-data/risingwave/pull/3143/commits/66eb59c4c0b8300469bd0421c5adf8303766b102.

**Pull request closed without merging**